Adds guzzle http client library

danielpieper opened the pull request:
Sends notifications using guzzle post requests

Because of strict php.ini settings (allow_url_fopen=0) we cannot use file_get_contents to send out airbrake notifications. So i replaced it with the guzzle library.

Commit:
Adds guzzle http client library
Sends notifications using guzzle post requests

kyrylo commented:
Thanks for the PR! I assume this is a fix for https://github.com/airbrake/phpbrake/issues/31. Looks like the build is failing, though.

danielpieper commented:
@kyrylo yes it's failing because the CircleCi uses an older version of PHP (5.4) but the guzzle library needs at least version 5.5.
Is it a constraint for you to be compatible with older versions of PHP or could you update the CircleCi PHP version?

vmihailenco commented:
Sorry for delay, but I think that we need to add support for several HTTP-clients and then provide users ability to switch between raw PHP client and guzzle. Then we can consider making guzzle a default backend.

vmihailenco commented on src/Notifier.php:
```diff
@@ -152,21 +153,9 @@ public function buildNotice($exc)
-            'headers' => isset($http_response_header) ? $http_response_header : [],
-            'data' => $respData,
-        ];
+        $handler = (isset($this->opt['httpClient'])?$this->opt['httpClient']:null);
```
Can you please add spaces around `?` and `:`?

vmihailenco commented on src/Notifier.php:
```diff
@@ -152,21 +153,9 @@ public function buildNotice($exc)
-            'data' => $respData,
-        ];
+        $handler = (isset($this->opt['httpClient'])?$this->opt['httpClient']:null);
+        $client = Http\Factory::createHttpClient($handler);
```
Can you move this to constructor so we don't create new client each time?

vmihailenco commented:
Looks good. Can you please address my comments and squash commits so I can merge this? Thanks.

vmihailenco commented on src/Http/Factory.php:
```diff
@@ -0,0 +1,65 @@
+
+namespace Airbrake\Http;
+
+use GuzzleHttp\Client;
```
Just wonder if this works when guzzle is not installed...

danielpieper committed:
Adds support for multiple http clients (default php, curl and guzzle)

danielpieper commented:
Thanks for your review, i fixed some issues but still tests and documentation/ readme are missing. i will add them soon

kyrylo commented:
Gentle ping @danielpieper. Would love to see this PR merged! :)

Commit:
Adds httpClient documentation
Removes (magic) client detection
Adds factory test
Disables guzzle test for php 5.4

danielpieper commented:
@kyrylo I removed the (magic) client detection to stick to the live client's behavior. As the guzzle client library requires PHP >= 5.5 this test will be skipped on circle ci.

kyrylo commented on README.md:
```diff
@@ -160,6 +160,24 @@ $notifier = new Airbrake\Notifier([
+- Curl needs the curl php extension installed. See phpinfo().
+- The default client uses the php function "file_get_contents". Make sure
+"allow_url_fopen" is set to "1" in your php.ini.
+If not set the default client is used.
```
This is pretty vague. Could we mention what exactly this 'default' client is?

danielpieper replied:
this line describes the default client. Which information would you like to add?

kyrylo replied:
I was thinking about linking to the file, but seems like `file_get_contents` is well-known. Never mind :)

kyrylo commented on src/Notifier.php:
```diff
@@ -51,6 +58,9 @@ public function __construct($opt = [])
             });
         }
+
+        $handler = (isset($this->opt['httpClient']) ? $this->opt['httpClient'] : null);
```
Couldn't we simply write:

``` php
$handler = $this->opt['httpClient'];
```

danielpieper replied:
this would trigger a notice:

```
php > $test = [];
php > echo $test['does-not-exist'];
PHP Notice:  Undefined index: does-not-exist in php shell code on line 1
```

In PHP7 we could use the null coalescing operator instead of isset()

kyrylo replied:
Thanks. We should stick to this code then, since we want to support more PHP versions.

kyrylo commented:
Thanks, @danielpieper! It looks fantastic. I just have a few questions/comments, but overall this looks good.
